File: typescript/parser/src/parser/json/jnumber-parser.ts
import { Parser } from '../common/parser'
import { JParser } from './jparser'
import { ParseResult, ParseSuccess, ParseFailer } from '../common/parser-result'
import * as P from '../common/parser'
import { JNumber } from './type/jnumber'
import { Some, None } from '../common/type/option'
import { Option } from '../common/type/option';

export class JNumberParser implements JParser<JNumber> {
	signP: Parser<Option<string>> = P.option(P.or(P.string("+"), P.string("-")))
	baseP = P.or(P.string("E"), P.string("e"))
	numberP: Parser<string> = P.or(P.or(P.or(P.or(P.or(P.or(P.or(P.or(P.or(P.string("1"), P.string("2")), P.string("3")), P.string("4")),
		P.string("5")), P.string("6")), P.string("7")), P.string("8")), P.string("9")), P.string("9"))

	intP: Parser<string[]> = P.plus(this.numberP)
	floatP = P.option(P.pair2(P.string("."), this.intP))

	parse(input: string): ParseResult<JNumber> {
		let next = input
		let numberStr = ""
		const signResult = this.signP.parse(next)
		if (signResult instanceof ParseFailer) {
			return new ParseFailer(signResult.message, signResult.next)
		} else {
			next = signResult.next
			if (signResult.value instanceof Some) {
				numberStr += signResult.value.value
			}
		}
		const numberResult = this.intP.parse(next)
		if (numberResult instanceof ParseFailer) {
			return new ParseFailer(numberResult.message, numberResult.next)
		} else {
			next = numberResult.next
			numberResult.value.forEach(n => {
				numberStr += n
			})
		}
		const floatResult: ParseResult<Option<[string, string[]]>> = this.floatP.parse(next)
		if (floatResult instanceof ParseFailer) {
			return new ParseFailer(floatResult.message, floatResult.next)
		} else {
			next = floatResult.next
			if (floatResult.value instanceof Some) {
				numberStr += floatResult.value.value[0]
				floatResult.value.value[1].forEach(i => {
					numberStr += i
				})
			}
		}

		const baseResult: ParseResult<Option<[string, Option<string>, string[]]>> = P.option(P.pair3(this.baseP, this.signP, this.intP)).parse(next)
		if(baseResult instanceof ParseFailer) {
			return new ParseFailer(baseResult.message, baseResult.next)
		} else {
			next = baseResult.next
			if(baseResult.value instanceof Some) {
				numberStr += baseResult.value.value[0]
				if(baseResult.value.value[1] instanceof Some) {
					numberStr += baseResult.value.value[1].value
				}
				baseResult.value.value[2].forEach(i => {
					numberStr += i
				})
			}
		}
		return new ParseSuccess(new JNumber(Number(numberStr)), next)
	}
}
export const jNumberParser = new JNumberParser()

File: scala/parsers/src/main/scala/jp/co/teruuu/parser/json/JNumberParser.scala
package jp.co.teruuu.parser.json

import jp.co.teruuu.parser.common._
import jp.co.teruuu.parser.json.`type`.{JNumber, JString}

object JNumberParser extends JParser[JNumber] {
  lazy val signP = Parser.string("+").or(Parser.string("-")).option()
  lazy val baseP = Parser.string("E").or(Parser.string("e"))
  lazy val ztonP = Parser.string("0").or(Parser.string("1")).or(Parser.string("2")).or(Parser.string("3")).
    or(Parser.string("4")).or(Parser.string("5")).or(Parser.string("6")).
    or(Parser.string("7")).or(Parser.string("8")).or(Parser.string("9"))
  lazy val parser1 = new Pair2(new Pair2(signP, ztonP.plus()), (new Pair2(Parser.string("."), ztonP.plus())).option())
  lazy val parser2 = (new Pair3(baseP, signP, ztonP.plus())).option()

  override def parse(input: String): ParseResult[JNumber] = {
    parser1.parse(input) match {
      case result1: ParseOk[((Option[String], List[String]), Option[(String, List[String])])] => {
        parser2.parse(result1.next) match {
          case result2: ParseOk[Option[(String, Option[String], List[String])]] => {
            var value = 0
            var deci = 0
            var base = 0
            val valueStr = result1.value._1._2.foldLeft(result1.value._1._1.getOrElse(""))((acc, cur) => {
              acc + cur
            })
            value = Integer.valueOf(valueStr)
            if(result1.value._2.isDefined) {
              val decStr = result1.value._2.get._2.foldLeft("")((acc, cur) => {
                acc + cur
              })
              deci = Integer.valueOf(decStr)
            }
            if(result2.value.isDefined) {
              val baseStr = result2.value.get._3.foldLeft(result2.value.get._2.getOrElse(""))((acc, cur) => {
                acc + cur
              })
              base = Integer.valueOf(baseStr)
            }
            ParseOk(JNumber(value, deci, base), result2.next)
          }
          case result2: ParseNg[_] => ParseNg(result2.message, result2.next)
        }
      }
      case result1: ParseNg[_] => ParseNg(result1.message, result1.next)
    }
  }
}
